### PR TITLE
Enable public MEC on CAPZ

### LINKED ISSUE #2660: Enable CAPZ to deploy cluster on Azure Public MEC
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

I'd like to make CAPZ able to deploy cluster on [Azure Public MEC](https://azure.microsoft.com/en-us/solutions/public-multi-access-edge-compute-mec).

Since Azure Public MEC share same Azure RP with public azure, and azure api is exactly the same, the expected change is little:
 - Add _extendedlocation_ parameter when creating azure resources
 - Don't set AvailablityZone (public MEC don't have this feature yet)
 - Use Loadblancer outbound instead of NAT Gateway (public MEC don't have NAT gateway)
 - Use Out of tree cloud controller manager. (public MEC support is only added to OOT)
 - Add related Test

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

My workmate and I will submit PR for this feature, but not familiar with the community process, so just want to raise a feature request to let community know.

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`): 

https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2660#tasklist-block-44e3d134-f6e2-4560-af5a-7c56b1890b3d

### PR DESCRIPTION
**Pull request by upxinxin**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

To make CAPZ able to deploy cluster on [Azure Public MEC](https://azure.microsoft.com/en-us/solutions/public-multi-access-edge-compute-mec). Changes we did:

1. Add extendedlocation in AzureCluster CRD

2. Passing extendedlocation info to azure SDK

3. Add a new flavor 'edgezone'. (remove nat-gateway, use OOT, add extendedlocation variables)

4. Disable availabilitySet and availabilityZone for public MEC

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #2660

**Special notes for your reviewer**:

Public MEC is not public accessible right now. To be able to create cluster on public MEC, an msft internal subscription with public MEC location access is needed. We've tested the feature e2e, but I think we may need a discussion on how to verify this feature from reviewer side.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

 

**TODOs**:

 

- [ ] squashed commits

- [ ] includes documentation

- [ ] adds unit tests

 

**Release note**:

```release-note
Enable CAPZ to deploy cluster on Azure Public MEC
```

### CODE REVIEW

File: azure/converters/extendedlocation.go
Comment (sonasingh46) [inline azure/converters/extendedlocation.go:48]: Adding a unit test for these functions will be useful.

File: azure/scope/cluster.go
Comment (sonasingh46) [inline azure/scope/cluster.go:205]: Do we need to add ExtendedLocation here? 
As per the issue it is mentioned that public MEC don't have NAT gateway

File: controllers/azurecluster_reconciler.go
Comment (sonasingh46) [inline controllers/azurecluster_reconciler.go:149]: Is it possible to have ExtendedLocation without name? If not, then the second condition is not required.

File: controllers/azuremachine_reconciler.go
Comment (sonasingh46) [inline controllers/azuremachine_reconciler.go:60]: Again, do we require NAT rules?

File: controllers/helpers.go
Comment (sonasingh46) [inline controllers/helpers.go:259]: Can we convert these to be methods of azure.ClusterScoper?

Comment (sonasingh46) [reply controllers/helpers.go:259]: Also, move this to azure/scope/cluster.go file. 

Example: 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/cluster.go#L699

**Comment by sonasingh46** [comment]
General questions: 
- For extended location, is the regular azure location relevant or simply it is not used? 
- Having an e2e for validation should be ideal but as it is mentioned that it is not publicly accessible, let us see what other folks have to say on this or if we have some way to navigate it.

File: api/v1beta1/types_class.go
Comment (CecileRobertMichon) [inline api/v1beta1/types_class.go:30]: In the case where `ExtendedLocation` is set, is `Location` also needed? If not, what would the user set it to since `Location`  is a required field?

File: templates/cluster-template-edgezone.yaml
Comment (mogliang) [reply templates/cluster-template-edgezone.yaml:1]: Yes, I think an e2e test to cover public MEC scenario is necessary. Will CI run the e2e test? if yes, we'll contact PM team and try to get approval to grant some public MEC sites to CI subscription.  

As regarding to the document, I think it's a bit too early at this moment. There isn't any public user can access public MEC yet, besides, there are some wording related to pubicMEC/EdgeZone not fully settled. We can create internal doc and share to users who need access public MEC, and will publish document when product is ready to public users.

File: azure/scope/cluster.go
Comment (upxinxin) [reply azure/scope/cluster.go:205]: Will remove ExtendedLocation in nodeNatGateway here, also in azureBastion as public MEC has not supported it.

File: azure/converters/extendedlocation.go
Comment (upxinxin) [reply azure/converters/extendedlocation.go:48]: Will add a unit test here

File: controllers/azurecluster_reconciler.go
Comment (upxinxin) [reply controllers/azurecluster_reconciler.go:149]: Only ExtendedLocation instance with name and type can go through the verification and go into reconciler. So will remove the second condition.

File: controllers/azuremachine_reconciler.go
Comment (upxinxin) [reply controllers/azuremachine_reconciler.go:60]: Will remove it

File: api/v1beta1/types_class.go
Comment (upxinxin) [reply api/v1beta1/types_class.go:30]: Yes, Location is needed. 
ExtendedLocation is attached to Location.

**Comment by upxinxin** [comment]
> General questions:
> 
> * For extended location, is the regular azure location relevant or simply it is not used?
> * Having an e2e for validation should be ideal but as it is mentioned that it is not publicly accessible, let us see what other folks have to say on this or if we have some way to navigate it.

1. ExtendedLocation is attached to azure location. Azure location matters.
2. Yes, we are preparing an e2e test to validate the feature on public MEC. But there still needs discussion for running test.

File: templates/cluster-template-edgezone.yaml
Comment (CecileRobertMichon) [reply templates/cluster-template-edgezone.yaml:1]: We shouldn't merge any code in this public repo that can't be used by users. If MEC is not ready to be used, why make this PR now? 

At the very least, there should be a doc explaining the feature is not ready yet so users don't get confused, and it should be behind a feature flag such that any user that tries to set ExtendedLocation without the feature flag being enabled gets a validation error.

Comment (mogliang) [reply templates/cluster-template-edgezone.yaml:1]: Thanks for the suggestion.

Public MEC is in public preview, and have customers now. We have another product (in private preview) which can also utilize this feature to deploy cluster. My understanding was inaccurate previously. We will add document to describe how to deploy on public MEC.

As refer to feature flag, I understand we need add here feature/feature.go and check it in controller right? Is there any additional thing or process need to be done when we add feature flag?

Comment (CecileRobertMichon) [reply templates/cluster-template-edgezone.yaml:1]: Sorry I missed this. 

> I understand we need add new flag in feature/feature.go and check it in controller right

Yes, that's exactly right. You can also add it to https://github.com/kubernetes-sigs/cluster-api/blob/main/config/manager/manager.yaml#L25 with a "false" value by default so it's easier to enable (via environment variable, when running clusterctl init). Here are some docs about it https://cluster-api.sigs.k8s.io/tasks/experimental-features/experimental-features.html.

And here is an example of returning a validation error in the admission webhook when the feature flag is not set: https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/api/v1beta1/machinepool_webhook.go#L103

File: controllers/helpers.go
Comment (upxinxin) [reply controllers/helpers.go]: Has convert to ExtendedLocationName() and ExtendedLocationType() in azure.ClusterScoper and move to azure/scope/cluster.go file.

File: templates/cluster-template-edgezone.yaml
Comment (upxinxin) [reply templates/cluster-template-edgezone.yaml:1]: Thanks for your detailed explanation.
I have added "EdgeZone" feature flag, set it with "false" value by default and add "EdgeZone" feature flag check in _azurecluster_webhook.go_.

**Comment by CecileRobertMichon** [comment]
/retest

apiversion upgrade test should be fixed since #2961 merged

**Comment by mboersma** [comment]
I think the e2e failure is a known issue when cleaning up a cluster, so there's just the linter complaining about formatting in feature.go.

**Comment by upxinxin** [comment]
> I think the e2e failure is a known issue when cleaning up a cluster, so there's just the linter complaining about formatting in feature.go.
Linter problem has been solved, and e2e test has succeeded this time luckily.

**Comment by upxinxin** [comment]
> I think the e2e failure is a known issue when cleaning up a cluster, so there's just the linter complaining about formatting in feature.go.

Linter problem has been solved, and e2e test has succeeded this time luckily.

File: azure/scope/cluster.go
Comment (mboersma) [inline azure/scope/cluster.go:744]: These are simple functions, but we should add unit tests in `cluster_test.go` to exercise them and use `ExtendedLocation`.

File: docs/book/src/topics/publicmec-clusters.md
Comment (mboersma) [inline docs/book/src/topics/publicmec-clusters.md:8]:
```suggestion
Cluster API Provider Azure (CAPZ) has experimental support for deploying clusters on [Azure Public MEC](https://azure.microsoft.com/en-us/solutions/public-multi-access-edge-compute-mec). Before you begin, you need an Azure subscription which has access to Public MEC.
```

File: docs/book/src/topics/publicmec-clusters.md
Comment (mboersma) [inline docs/book/src/topics/publicmec-clusters.md:10]:
```suggestion
To deploy a cluster on Public MEC, provide extended location info through environment variables and use the "edgezone" flavor.
```

File: docs/book/src/topics/publicmec-clusters.md
Comment (mboersma) [inline docs/book/src/topics/publicmec-clusters.md:14]:
```suggestion
The clusterctl "edgezone" flavor exists to deploy clusters on Public MEC. This flavor requires the following environment variables to be set before executing `clusterctl`.
```

File: docs/book/src/topics/publicmec-clusters.md
Comment (mboersma) [inline docs/book/src/topics/publicmec-clusters.md:34]:
```suggestion
Export the following variables to your current shell.
```

File: docs/book/src/topics/publicmec-clusters.md
Comment (mboersma) [inline docs/book/src/topics/publicmec-clusters.md:85]:
```suggestion
To fix the "False" READY status, [Azure cloud provider components](https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/helm/cloud-provider-azure) need to be installed by Helm.
```

**Review by mboersma — COMMENTED** [review]
This is looking good overall, I had some minor comments.

File: azure/scope/cluster.go
Comment (upxinxin) [reply azure/scope/cluster.go:744]: Has added unit tests for ExtendedLocationName() and ExtendedLocationType().

**Comment by CecileRobertMichon** [comment]
/retest

AKS test should be fixed now

**Comment by CecileRobertMichon** [comment]
@upxinxin can you please squash commits?

/hold

File: api/v1beta1/azurecluster_webhook.go
Comment (CecileRobertMichon) [inline api/v1beta1/azurecluster_webhook.go:53]: why is this not part of the `validateCluster` function?

File: api/v1beta1/types_class.go
Comment (CecileRobertMichon) [inline api/v1beta1/types_class.go:28]:
```suggestion
	// ExtendedLocation is an optional set of ExtendedLocation properties for clusters in Azure public MEC.
```

File: api/v1beta1/types_class.go
Comment (CecileRobertMichon) [inline api/v1beta1/types_class.go:64]:
```suggestion
	// Type defines a type for extended location (eg. EdgeZone).
```

Comment (CecileRobertMichon) [reply api/v1beta1/types_class.go:64]: is there a well known set of valid type values? If so we should validate this with a kubebuilder Enum

File: azure/converters/extendedlocation.go
Comment (CecileRobertMichon) [inline azure/converters/extendedlocation.go:2]:
```suggestion
Copyright 2022 The Kubernetes Authors.
```

File: controllers/azuremachine_reconciler.go
Comment (CecileRobertMichon) [inline controllers/azuremachine_reconciler.go:55]: we shouldn't need to do this, we can simply make the availability set service reconcile a no-op when `ExtendedLocation` is enabled, by returning a nil spec in machine.go. This should already be the case with https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2836/files#diff-b6931ac61f37a13f66fc7836e039bf714a0553d5db4f313cc8ef31ca5f5647b9R498

We can't remove services that are unused based on every Machine configuration that doesn't need them in here otherwise the code would become a mess of if/else statements.

File: controllers/helpers.go
Comment (CecileRobertMichon) [inline controllers/helpers.go:256]: is this supported in all versions of cloud provider both in tree and out of tree? If not what's the minimum version that supports it?

Comment (upxinxin) [reply controllers/helpers.go:256]: It is not supported in "in-tree" cloud provider, but supported in "out-of-tree" cloud provider since version v1.0.3.

File: controllers/azuremachine_reconciler.go
Comment (upxinxin) [reply controllers/azuremachine_reconciler.go:55]: Yes, we don't need the if statement as nil spec in machine.go will help return nil in availability set reconcile func.

Thanks for your kind and helpful comment.

File: api/v1beta1/types_class.go
Comment (upxinxin) [reply api/v1beta1/types_class.go]: Yes, I have added the Enum for type value.

File: api/v1beta1/azurecluster_webhook.go
Comment (upxinxin) [reply api/v1beta1/azurecluster_webhook.go]: It has been moved into `validateCluster` function.

**Comment by upxinxin** [comment]
>


> @upxinxin can you please squash commits?
> 
> /hold

Sure. I have squashed commits. Please help review it.

**Comment by upxinxin** [comment]
Extra "force-push" operations are to fix new branch conflicts and codeshell problems.

File: azure/scope/machine.go
Comment (CecileRobertMichon) [inline azure/scope/machine.go:497]: can you please add a short code comment about why `ExtendedLocation` disabled AvailabilitySets for future reference?

File: hack/util.sh
Comment (CecileRobertMichon) [inline hack/util.sh:56]: is this the only supported region currently?

**Comment by upxinxin** [comment]
Public MEC cluster creation failed in "pull-cluster-api-provider-azure-e2e-optional" test because CAPZ CI subscription has no access to deploy clusters on extended location. 

If you want to test it by your own, we can contact our PM team and try to get approval to grant some public MEC sites to CI subscription. Or, we can help maintain public MEC related feature.

File: hack/util.sh
Comment (upxinxin) [reply hack/util.sh:56]: Yes, we have not replicated CAPI image to other edgezone region. But we will support more regions in the future.

File: azure/scope/machine.go
Comment (upxinxin) [reply azure/scope/machine.go:497]: Sure.

**Comment by upxinxin** [comment]
It seems like a short code comment change has made LGTM label removed.

Could you please help add LGTM label again? @CecileRobertMichon

**Comment by CecileRobertMichon** [comment]
> Public MEC cluster creation failed in "pull-cluster-api-provider-azure-e2e-optional" test because CAPZ CI subscription has no access to deploy clusters on extended location.
>
> If you want to test it by your own, we can contact our PM team and try to get approval to grant some public MEC sites to CI subscription. Or, we can help maintain public MEC related feature.

/hold

we should not be adding a test in the CI test suite that can't be run by our CI sub as it will break our periodic jobs

> Or, we can help maintain public MEC related feature

can you please expand on what you mean by this ^ ?

**Comment by mogliang** [comment]
> can you please expand on what you mean by this ^ ?

upxinxin and I are from edgezone team, and would like to continue contributing on Kubernetes community, e.g. maintain edgezone features on CAPZ repo.

> we should not be adding a test in the CI test suite that can't be run by our CI sub as it will break our periodic jobs

The test failure is due to CI sub don't have access to edgezone site resource, let me check with PM on this.

**Comment by upxinxin** [comment]
@CecileRobertMichon We have passed e2e-optional test. Please help review it again.

File: azure/converters/extendedlocation.go
Comment (sonasingh46) [inline azure/converters/extendedlocation.go:30]: nit, gosimple: The else block is not needed here.

File: azure/converters/extendedlocation.go
Comment (sonasingh46) [inline azure/converters/extendedlocation.go:42]: nit, gosimple: The else block is not needed here.

**Review by sonasingh46 — APPROVED** [review]
one minor nit. 
Otherwise 
/lgtm

File: azure/converters/extendedlocation.go
Comment (upxinxin) [reply azure/converters/extendedlocation.go]: Has removed it.

File: azure/converters/extendedlocation.go
Comment (upxinxin) [reply azure/converters/extendedlocation.go]: Has removed it.

**Review by sonasingh46 — APPROVED** [review]
/lgtm

**Comment by upxinxin** [comment]
Serveral sub-tests in e2e-optional test failed with the error "cannot create more than 100 public IP addresses for this subscription in this region". It seems like public IPs in your CI test sub needs to be cleaned. @CecileRobertMichon

**Comment by CecileRobertMichon** [comment]
yes we ran into quota issues in the subscription yesterday https://kubernetes.slack.com/archives/CEX9HENG7/p1677709131827159, should be fixed now

/retest
/approve

**Pull request merged**
